File: backend/apps/audits/finding.py
"""Parse labelled recommendation text written by the create-recommendation wizard.

Headings are part of the stored `Recommendation.text` contract — keep them in
sync with `frontend/lib/finding.ts`. Changing a heading breaks old records.
"""
from __future__ import annotations

import re
from typing import TypedDict

HEADINGS = {
    "title": "العنوان",
    "condition": "الوضع القائم",
    "criteria": "المعيار",
    "effect": "الأثر",
    "statement": "التوصية",
    "objective": "الهدف",
    "required_action": "الإجراء المطلوب",
    "outcome": "النتيجة المتوقعة",
    "success": "مؤشر النجاح",
    "evidence": "الأدلة المتوقعة",
    "notes": "الملاحظات",
}

HEADING_ALIASES = {
    "الملاحظة": "condition",
    "المخاطر": "effect",
    "Finding": "condition",
    "Objective": "objective",
    "Risks": "effect",
    "Recommendation": "statement",
    "Success Indicator": "success",
}

SECTION_ORDER = tuple(HEADINGS.keys())

_HEADING_LOOKUP: dict[str, str] = {heading: sid for sid, heading in HEADINGS.items()}
_HEADING_LOOKUP.update(HEADING_ALIASES)

_HEADING_PATTERN = "|".join(
    re.escape(name) for name in [*HEADINGS.values(), *HEADING_ALIASES.keys()]
)
_HEADING_LINE = re.compile(rf"^({_HEADING_PATTERN}):\s*$")

# Words that appear in every wizard record; they must not drive similarity.
_STRUCTURAL_STOP = {
    "العنوان", "الوضع", "القائم", "المعيار", "الاثر", "الأثر", "التوصيه", "التوصية",
    "الهدف", "الاجراء", "الإجراء", "المطلوب", "النتيجه", "النتيجة", "المتوقعه",
    "المتوقعة", "مؤشر", "النجاح", "الادله", "الأدلة", "الملاحظات", "title",
    "condition", "criteria", "finding", "recommendation", "objective",
}


class ParsedFinding(TypedDict):
    sections: dict[str, str]
    preamble: str
    structured: bool


def parse_finding(text: str | None) -> ParsedFinding:
    value = (text or "").strip()
    if not value:
        return {"sections": {}, "preamble": "", "structured": False}

    sections: dict[str, str] = {}
    preamble: list[str] = []
    current: str | None = None
    buffer: list[str] = []

    def flush() -> None:
        nonlocal buffer, current
        if current:
            body = "\n".join(buffer).strip()
            if body:
                sections[current] = body
        buffer = []

    for line in value.splitlines():
        match = _HEADING_LINE.match(line.strip())
        if match:
            flush()
            current = _HEADING_LOOKUP[match.group(1)]
            continue
        if current:
            buffer.append(line)
        else:
            preamble.append(line)
    flush()

    return {
        "sections": sections,
        "preamble": "\n".join(preamble).strip(),
        "structured": bool(sections),
    }


def section_body(text: str | None, section_id: str) -> str:
    return parse_finding(text)["sections"].get(section_id, "").strip()


def clip(value: str | None, limit: int = 220) -> str:
    clean = re.sub(r"\s+", " ", (value or "").strip())
    if len(clean) <= limit:
        return clean
    return clean[: limit].rstrip() + "…"


def case_title(text: str | None, max_len: int = 90) -> str:
    parsed = parse_finding(text)
    source = (
        parsed["sections"].get("title")
        or parsed["sections"].get("condition")
        or parsed["preamble"]
        or parsed["sections"].get("statement")
        or (text or "")
    )
    first = next((line.strip() for line in source.splitlines() if line.strip()), "")
    return clip(first, max_len)


def brief_excerpt(text: str | None, limit: int = 180) -> str:
    """One-line snippet for lists and similarity cards — never the full dump."""
    parsed = parse_finding(text)
    for key in ("title", "condition", "statement"):
        body = parsed["sections"].get(key)
        if body:
            return clip(body.splitlines()[0], limit)
    if parsed["preamble"]:
        return clip(parsed["preamble"].splitlines()[0], limit)
    return clip(text, limit)


def semantic_payload(text: str | None, root_cause: str | None = None) -> str:
    """Heading-free text used for embeddings so template labels do not match."""
    parsed = parse_finding(text)
    parts: list[str] = []
    if parsed["structured"]:
        for key in (
            "title",
            "condition",
            "criteria",
            "effect",
            "statement",
            "objective",
            "required_action",
            "root_cause",
        ):
            body = parsed["sections"].get(key)
            if body:
                parts.append(body)
        if parsed["preamble"]:
            parts.insert(0, parsed["preamble"])
    else:
        if parsed["preamble"]:
            parts.append(parsed["preamble"])
        elif text:
            parts.append(text)
    if (root_cause or "").strip():
        parts.append(root_cause.strip())
    payload = "\n".join(p.strip() for p in parts if p and p.strip())
    return payload.strip() or (text or "").strip()


def is_repeated_placeholder(text: str | None) -> bool:
    """True when several labelled sections repeat the same short token."""
    bodies = [body.strip() for body in parse_finding(text)["sections"].values() if body.strip()]
    if len(bodies) < 3:
        return False
    unique = {re.sub(r"\s+", " ", b) for b in bodies}
    if len(unique) != 1:
        return False
    only = next(iter(unique))
    return len(only) < 48


def filled_sections(text: str | None) -> list[str]:
    parsed = parse_finding(text)
    return [key for key in SECTION_ORDER if parsed["sections"].get(key, "").strip()]


def structural_tokens() -> set[str]:
    return set(_STRUCTURAL_STOP)

File: frontend/lib/finding.ts
/**
 * The backend stores a recommendation as free text (`Recommendation.text`) plus
 * a separate `root_cause` field — there are no per-section columns.
 *
 * To keep audit working-paper structure without inventing backend fields, the
 * composer writes labelled sections into that single text field and the
 * workspace parses them back out. Text authored elsewhere (or by older records)
 * simply has no sections and renders as plain prose.
 */

import { currentT } from "./i18n/messages";

export type FindingSectionId =
  | "title"
  | "condition"
  | "criteria"
  | "effect"
  | "statement"
  | "objective"
  | "required_action"
  | "outcome"
  | "success"
  | "evidence"
  | "notes";

export interface FindingSectionDef {
  id: FindingSectionId;
  /** Heading written into `text`. Changing this breaks parsing of old records. */
  heading: string;
  label: string;
  hint: string;
  required: boolean;
  rows: number;
}

export const FINDING_SECTIONS: FindingSectionDef[] = [
  {
    id: "title",
    heading: "العنوان",
    label: "عنوان الملاحظة",
    hint: "جملة واحدة تصف جوهر الملاحظة، تظهر في السجل ورأس الملف.",
    required: false,
    rows: 1,
  },
  {
    id: "condition",
    heading: "الوضع القائم",
    label: "الوضع القائم — ما الذي وُجد؟",
    hint: "صف الواقع الذي لوحظ أثناء التدقيق بدقة وبلا أحكام.",
    required: true,
    rows: 4,
  },
  {
    id: "criteria",
    heading: "المعيار",
    label: "المعيار — ما الذي كان يجب أن يحدث؟",
    hint: "النظام أو التعليمات أو الإجراء المعتمد الذي تمت المقارنة به.",
    required: false,
    rows: 3,
  },
  {
    id: "effect",
    heading: "الأثر",
    label: "الأثر — ما النتيجة المترتبة؟",
    hint: "الأثر المالي أو التشغيلي أو الرقابي أو السمعي المترتب على الوضع القائم.",
    required: false,
    rows: 3,
  },
  {
    id: "statement",
    heading: "التوصية",
    label: "نص التوصية",
    hint: "صياغة واضحة وقابلة للتنفيذ والقياس تحدد ما يجب أن تفعله الدائرة.",
    required: true,
    rows: 4,
  },
  {
    id: "objective",
    heading: "الهدف",
    label: "الهدف من التوصية",
    hint: "ما الذي يجب أن يتحقق لدى الإدارة بعد التنفيذ؟",
    required: false,
    rows: 3,
  },
  {
    id: "required_action",
    heading: "الإجراء المطلوب",
    label: "الإجراء المطلوب من الإدارة",
    hint: "ما الذي يُطلب من الدائرة فعله، بصيغة قابلة للتنفيذ.",
    required: false,
    rows: 3,
  },
  {
    id: "outcome",
    heading: "النتيجة المتوقعة",
    label: "النتيجة التصحيحية المتوقعة",
    hint: "ما الذي سيتغيّر عند تنفيذ التوصية؟",
    required: false,
    rows: 3,
  },
  {
    id: "success",
    heading: "مؤشر النجاح",
    label: "مؤشر النجاح",
    hint: "كيف ستعرف الرقابة أن المعالجة تحققت؟",
    required: false,
    rows: 2,
  },
  {
    id: "evidence",
    heading: "الأدلة المتوقعة",
    label: "الأدلة المتوقعة لإثبات التنفيذ",
    hint: "يوجّه الدائرة إلى ما يجب رفعه لاحقاً كدليل على التنفيذ.",
    required: false,
    rows: 3,
  },
  {
    id: "notes",
    heading: "الملاحظات",
    label: "ملاحظات إضافية",
    hint: "أي سياق إضافي لا يندرج تحت الأقسام الأخرى.",
    required: false,
    rows: 3,
  },
];

/** Extra headings recognised when parsing older or free-typed text. */
const HEADING_ALIASES: Record<string, FindingSectionId> = {
  "الملاحظة": "condition",
  "المخاطر": "effect",
  Finding: "condition",
  Objective: "objective",
  Risks: "effect",
  Recommendation: "statement",
  "Success Indicator": "success",
};

export function sectionUi(id: FindingSectionId): { label: string; hint: string } {
  const T = currentT();
  const map: Record<FindingSectionId, { label: string; hint: string }> = {
    title: { label: T.create.findingTitle, hint: T.create.findingTitleHint },
    condition: { label: T.create.condition, hint: T.create.conditionHint },
    criteria: { label: T.create.criteria, hint: T.create.criteriaHint },
    effect: { label: T.create.effect, hint: T.create.effectHint },
    statement: { label: T.create.statement, hint: T.create.statementHint },
    objective: { label: T.create.objectiveLabel, hint: T.create.objectiveHint },
    required_action: { label: T.create.requiredActionLabel, hint: T.create.requiredActionHint },
    outcome: { label: T.create.expectedOutcome, hint: T.create.expectedOutcomeHint },
    success: { label: T.create.successIndicator, hint: T.create.successHint },
    evidence: { label: T.create.expectedEvidence, hint: T.create.expectedEvidenceHint },
    notes: { label: T.create.extraNotes, hint: T.create.extraNotesHint },
  };
  return map[id];
}

export function localizedFindingSections(): FindingSectionDef[] {
  return FINDING_SECTIONS.map((section) => ({ ...section, ...sectionUi(section.id) }));
}

export type FindingDraft = Record<FindingSectionId, string>;

export const EMPTY_FINDING: FindingDraft = {
  title: "",
  condition: "",
  criteria: "",
  effect: "",
  statement: "",
  objective: "",
  required_action: "",
  outcome: "",
  success: "",
  evidence: "",
  notes: "",
};

/** Builds the single `text` value stored on the recommendation. */
export function composeFinding(draft: FindingDraft): string {
  return FINDING_SECTIONS.filter((section) => draft[section.id]?.trim())
    .map((section) => `${section.heading}:\n${draft[section.id].trim()}`)
    .join("\n\n");
}

export interface ParsedFinding {
  /** Sections found in the stored text, in canonical order. */
  sections: Array<{ id: FindingSectionId; heading: string; label: string; body: string }>;
  /** Text that appeared before any recognised heading, or the whole value. */
  preamble: string;
  structured: boolean;
}

const HEADING_LOOKUP = new Map<string, FindingSectionDef>(
  FINDING_SECTIONS.map((section) => [section.heading, section])
);
for (const [alias, id] of Object.entries(HEADING_ALIASES)) {
  const def = FINDING_SECTIONS.find((section) => section.id === id);
  if (def) HEADING_LOOKUP.set(alias, def);
}

const HEADING_PATTERN = [...FINDING_SECTIONS.map((section) => section.heading), ...Object.keys(HEADING_ALIASES)]
  .map((heading) => heading.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"))
  .join("|");

/** Splits stored text back into sections. Unstructured text returns as preamble. */
export function parseFinding(text: string): ParsedFinding {
  const value = (text ?? "").trim();
  if (!value) return { sections: [], preamble: "", structured: false };

  const pattern = new RegExp(`^(${HEADING_PATTERN}):\\s*$`, "u");
  const lines = value.split(/\r?\n/);

  const found: ParsedFinding["sections"] = [];
  const preamble: string[] = [];
  let currentId: FindingSectionId | null = null;
  let buffer: string[] = [];

  const flush = () => {
    if (currentId) {
      const def = FINDING_SECTIONS.find((section) => section.id === currentId)!;
      const body = buffer.join("\n").trim();
      if (body) found.push({ id: def.id, heading: def.heading, label: sectionUi(def.id).label, body });
    }
    buffer = [];
  };

  for (const line of lines) {
    const match = pattern.exec(line.trim());
    if (match) {
      flush();
      currentId = HEADING_LOOKUP.get(match[1])!.id;
      continue;
    }
    if (currentId) buffer.push(line);
    else preamble.push(line);
  }
  flush();

  return {
    sections: found,
    preamble: preamble.join("\n").trim(),
    structured: found.length > 0,
  };
}

/** Short one-line title for headers, tables and search results. */
export function caseTitle(text: string, max = 90): string {
  const parsed = parseFinding(text);
  const source =
    parsed.sections.find((section) => section.id === "title")?.body ||
    parsed.sections.find((section) => section.id === "condition")?.body ||
    parsed.preamble ||
    parsed.sections[0]?.body ||
    text ||
    "";
  const firstLine = source.split(/\r?\n/).find((line) => line.trim()) ?? "";
  const clean = firstLine.trim();
  return clean.length > max ? `${clean.slice(0, max).trimEnd()}…` : clean;
}
